### PR TITLE
Metrics for database, commit_log, bootstrapping, persistence

### PR DESCRIPTION
**Pull request by jra3**
Summary:
Adding some key metrics to the database component of m3db

Test Plan: run on one prod node

CC: @xichen2020 @robskillington

### CODE REVIEW

File: storage/database.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ import (
 	xio "github.com/m3db/m3db/x/io"
 	"github.com/m3db/m3x/errors"
 	"github.com/m3db/m3x/time"
+	"github.com/uber-go/tally"
```
Comment (xichen2020) [inline storage/database.go]: Insert a blank line before this import?

File: storage/database.go [hunk 2]
```diff
@@ -196,15 +248,21 @@ func (d *db) Write(
...
 
 	if !exists {
+		d.metrics.write.error.Inc(1)
+		d.metrics.write.latency.Stop(sw)
 		return fmt.Errorf("no such namespace %s", namespace)
 	}
 
-	return n.Write(ctx, id, timestamp, value, unit, annotation)
+	err := n.Write(ctx, id, timestamp, value, unit, annotation)
+	d.metrics.write.success.Inc(1)
```
Comment (xichen2020) [inline storage/database.go]: We should check the error to see if it actually succeeded

File: storage/database.go [hunk 3]
```diff
@@ -213,11 +271,19 @@ func (d *db) ReadEncoded(
...
+
 	if err != nil {
+		d.metrics.read.error.Inc(1)
+		d.metrics.read.latency.Stop(sw)
 		return nil, err
 	}
-	return n.ReadEncoded(ctx, id, start, end)
+
+	res, err := n.ReadEncoded(ctx, id, start, end)
+	d.metrics.read.success.Inc(1)
```
Comment (xichen2020) [inline storage/database.go]: Should check the error to see if it's a successful read

File: storage/database.go [hunk 4]
```diff
@@ -227,11 +293,20 @@ func (d *db) FetchBlocks(
...
-		return nil, xerrors.NewInvalidParamsError(err)
+		res := xerrors.NewInvalidParamsError(err)
+		d.metrics.fetchBlocks.error.Inc(1)
+		d.metrics.fetchBlocks.latency.Stop(sw)
+		return nil, res
 	}
-	return n.FetchBlocks(ctx, shardID, id, starts)
+
+	res, err := n.FetchBlocks(ctx, shardID, id, starts)
+	d.metrics.fetchBlocks.success.Inc(1)
```
Comment (xichen2020) [inline storage/database.go]: Should check the error to see if it's actually suceeded

File: storage/database.go [hunk 5]
```diff
@@ -290,6 +374,24 @@ func (d *db) getOwnedNamespaces() []databaseNamespace {
 	return namespaces
 }
 
+func (d *db) reportLoop() {
+	t := time.Tick(time.Second) // TODO: Make this an option, what is a good default
```
Comment (xichen2020) [inline storage/database.go]: Might wanna make this part of the database options?

File: instrument/options.go [hunk 6]
```diff
@@ -57,3 +61,7 @@ func (o *options) MetricsScope(value tally.Scope) Options {
 func (o *options) GetMetricsScope() tally.Scope {
 	return o.scope
 }
+
+func (o *options) GetGaugeInterval() time.Duration {
```
Comment (xichen2020) [inline instrument/options.go]: Probably also want a setter here so we can override it if we want

File: instrument/types.go [hunk 7]
```diff
@@ -38,4 +40,32 @@ type Options interface {
...
+
+// MethodMetrics is a bundle of common metrics with a uniform naming scheme
+type MethodMetrics struct {
+	Error   tally.Counter
+	Success tally.Counter
+	Latency tally.Timer
+}
+
+// ReportSuccessOrFailure increments Error/Success counter dependant on the error
+func (m *MethodMetrics) ReportSuccessOrFailure(e error) {
```
Comment (xichen2020) [inline instrument/types.go]: Move the implementation to a separate file? We usually only keep type defs in `types.go`

File: storage/database.go [hunk 8]
```diff
@@ -375,7 +363,8 @@ func (d *db) getOwnedNamespaces() []databaseNamespace {
 }
 
 func (d *db) reportLoop() {
-	t := time.Tick(time.Second) // TODO: Make this an option, what is a good default
+	interval := d.opts.GetInstrumentOptions().GetGaugeInterval()
```
Comment (xichen2020) [inline storage/database.go]: Perhaps call it `GetReportInterval` instead in case we want to add more metrics to the `reportLoop` method?

**Comment by robskillington** [comment]
LGTM

**Pull request merged**